realm/realm-library/src/main/java/io/realm/BaseRealm.java

```diff
@@ -637,47 +534,61 @@ protected static synchronized void validateAgainstExistingConfigurations(RealmCo
+                            "Remember to close() all the instances of the Realm before deleting its file.");
+                }
 
-        boolean realmDeleted = true;
-        String canonicalPath = configuration.getPath();
-        File realmFolder = configuration.getRealmFolder();
-        String realmFileName = configuration.getRealmFileName();
-        List<File> filesToDelete = Arrays.asList(new File(canonicalPath),
-                new File(realmFolder, realmFileName + ".lock"),
-                new File(realmFolder, realmFileName + ".lock_a"),
-                new File(realmFolder, realmFileName + ".lock_b"),
-                new File(realmFolder, realmFileName + ".log"));
-        for (File fileToDelete : filesToDelete) {
-            if (fileToDelete.exists()) {
-                boolean deleteResult = fileToDelete.delete();
-                if (!deleteResult) {
-                    realmDeleted = false;
-                    RealmLog.w("Could not delete the file " + fileToDelete);
+                String canonicalPath = configuration.getPath();
+                File realmFolder = configuration.getRealmFolder();
+                String realmFileName = configuration.getRealmFileName();
+                List<File> filesToDelete = Arrays.asList(new File(canonicalPath),
+                        new File(realmFolder, realmFileName + ".lock"),
+                        new File(realmFolder, realmFileName + ".lock_a"),
+                        new File(realmFolder, realmFileName + ".lock_b"),
+                        new File(realmFolder, realmFileName + ".log"));
+                for (File fileToDelete : filesToDelete) {
+                    if (fileToDelete.exists()) {
+                        boolean deleteResult = fileToDelete.delete();
+                        if (!deleteResult) {
+                            realmDeleted.set(false);
+                            RealmLog.w("Could not delete the file " + fileToDelete);
+                        }
+                    }
                 }
             }
-        }
+        });
 
-        return realmDeleted;
+        return realmDeleted.get();
     }
 
     /**
      * Compacts the Realm file defined by the given configuration.
      */
-    public static synchronized boolean compactRealm(RealmConfiguration configuration) {
+    public static boolean compactRealm(final RealmConfiguration configuration) {
         if (configuration.getEncryptionKey() != null) {
             throw new IllegalArgumentException("Cannot currently compact an encrypted Realm.");
         }
 
-        if (isFileOpen(configuration)) {
-            throw new IllegalStateException("Cannot compact an open Realm");
-        }
+        final AtomicBoolean result = new AtomicBoolean(false);
+
+        RealmCache.invokeWithGlobalRefCount(configuration, new RealmCache.Callback() {
+            @Override
+            public void onResult(int count) {
```

Comment:
if you need the following code to run in a synchronized block, just use an intrinsic or Reentrant lock